saleor/discount/utils/order.py

```diff
@@ -48,19 +53,56 @@ def create_or_update_discount_objects_from_promotion_for_order(
     lines_info: list["EditableOrderLineInfo"],
     database_connection_name: str = settings.DATABASE_CONNECTION_DEFAULT_NAME,
 ):
-    create_order_line_discount_objects_for_catalogue_promotions(lines_info)
-    # base unit price must reflect all actual catalogue discounts
+    delete_order_line_discount_objects_for_catalogue_promotions(lines_info)
+    # Update base price with catalogue discount reward. This price will be further used:
+    # - to qualify order for order promotion
+    # - to apply vouchers
     _update_base_unit_price_amount_for_catalogue_promotion(lines_info)
     create_order_discount_objects_for_order_promotions(
         order, lines_info, database_connection_name=database_connection_name
     )
 
 
-def create_order_line_discount_objects_for_catalogue_promotions(
+def delete_order_line_discount_objects_for_catalogue_promotions(
     lines_info: list["EditableOrderLineInfo"],
 ):
-    discount_data = prepare_line_discount_objects_for_catalogue_promotions(lines_info)
-    create_order_line_discount_objects(lines_info, discount_data)
+    """Delete catalogue discounts if other conflicting discounts exist.
+
+    - manual line discounts do not stuck with catalogue discounts
+    - gift line should not have any catalogue discounts associated
+    """
+    line_discounts_to_remove: list[OrderLineDiscount] = []
+
+    if not lines_info:
+        return
+
+    for line_info in lines_info:
+        line = line_info.line
+
+        # get the existing catalogue discount for the line
+        if existing_discounts := line_info.get_catalogue_discounts():
+            # Line should never have multiple catalogue discounts associated. Before
+            # introducing unique_type on discount models, there was such a possibility.
+            line_discounts_to_remove.extend(existing_discounts[1:])
```

Comment:
This code was introduced before. I just split the function for order and checkout. 

As Maciek suggested, this can be probably deleted and do the all deletions directly in manual line discount and gift logic.